### PR TITLE
Convert onboarding to Lit

### PR DESCRIPTION
**Pull request by balloob**
- Convert onboarding to Lit
- Onboarding will now return an auth code, so we can redirect to the main page directly.

onboarding-improv

Requires https://github.com/home-assistant/home-assistant/pull/21777

### CODE REVIEW

**Comment by awarecan** [comment]
We should allow password show clear text here (even set it as default). Because we don't have password_confirm field, if user either forget or have a typo in the password, he/she will have to redo on-boarding again.

**Comment by balloob** [comment]
That's a good point. We should do that for a future PR . Now that it's Lit, things will be easier.

File: src/onboarding/ha-onboarding.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+import { genClientId } from "home-assistant-js-websocket";
+import { litLocalizeLiteMixin } from "../mixins/lit-localize-lite-mixin";
+import { OnboardingStep, onboardUserStep } from "../data/onboarding";
+import { PolymerChangedEvent } from "../polymer-types";
+
+@customElement("ha-onboarding")
+class HaOnboarding extends litLocalizeLiteMixin(LitElement) {
+  public translationFragment = "page-onboarding";
+
+  @property() private _name = "";
```
Comment (iantrich) [inline src/onboarding/ha-onboarding.ts]: are these not initialized if you set the type to string?

File: src/onboarding/ha-onboarding.ts [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        value=${this._password}
+        @value-changed=${this._handleValueChanged}
+        required
+        type='password'
+        auto-validate
+        .errorMessage="${this.localize(
+          "ui.panel.page-onboarding.user.required_field"
+        )}"
+      ></paper-input>
+
```
Comment (iantrich) [inline src/onboarding/ha-onboarding.ts]: Add a password confirmation input and require the two fields to match

File: src/onboarding/ha-onboarding.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+      if (steps.every((step) => step.done)) {
+        // Onboarding is done!
+        document.location.href = "/";
+      }
+    } catch (err) {
+      alert("Something went wrong loading loading onboarding, try refreshing");
+    }
+  }
+
+  private _handleValueChanged(ev: PolymerChangedEvent<string>) {
```
Comment (iantrich) [inline src/onboarding/ha-onboarding.ts]:
```suggestion
  private _handleValueChanged(ev: PolymerChangedEvent<string>): void {
```

File: src/onboarding/ha-onboarding.ts [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    const name = (ev.target as any).name;
+    this[`_${name}`] = ev.detail.value;
+  }
+
+  private _maybePopulateUsername() {
```
Comment (iantrich) [inline src/onboarding/ha-onboarding.ts]:
```suggestion
  private _maybePopulateUsername(): void {
```

File: src/onboarding/ha-onboarding.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    const parts = this._name.split(" ");
+
+    if (parts.length) {
+      this._username = parts[0].toLowerCase();
+    }
+  }
+
+  private async _submitForm() {
```
Comment (iantrich) [inline src/onboarding/ha-onboarding.ts]:
```suggestion
  private async _submitForm(): Promise<void> {
```

File: src/onboarding/ha-onboarding.ts (hunk 1 above)
Comment (balloob) [reply src/onboarding/ha-onboarding.ts]: nope, it's undefined

File: src/onboarding/ha-onboarding.ts (hunk 2 above)
Comment (balloob) [reply src/onboarding/ha-onboarding.ts]: Done

**Review by iantrich — APPROVED** [review]

**Comment by balloob** [comment]
I've added a password confirm.

**Pull request merged**